[Refactor] Fixing mypy issues for `syft/service/` and `syft/util/`

khoaguin opened the pull request:
Fix mypy issues for `syft/service` and `syft/util` directories

khoaguin committed:
[refactor] done fixing mypy issues for `syft/service/request`

khoaguin committed:
[refactor] checking if `attr` has both `fget` and `fset` methods,
then call the methods instead of using `assert` since assert
statements are removed when compiling Python code to optimized
bytecode, causing security issue

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-request

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-request

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-request

kiendang committed:
Merge branch 'dev' into fix-mypy-issues-service-dir

khoaguin committed:
[refactor] done fixing mypy issues for `syft/service/code_history`

khoaguin committed:
[refactor] done fixing mypy issues for `service/project/project_service.py`

khoaguin committed:
[refactor] done fixing mypy issues for `syft/util/`

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-dir

khoaguin committed:
[refactor] done fixing mypy issues for `syft/service/action`

khoaguin committed:
[tests] skip `test_transfer_request_nonblocking` on windows
due to flakiness

kiendang commented on packages/syft/tests/syft/request/request_multiple_nodes_test.py:
```diff
@@ -149,6 +150,7 @@ def compute_sum(data) -> float:
 
 
 @pytest.mark.flaky(reruns=2, reruns_delay=1)
+@pytest.mark.skipif(sys.platform == "win32", reason="very flaky on windows")
```
This should be in a separate PR. You can create a separate PR for this and post it in the engineering channel to have others' opinion.

khoaguin replied:
Thanks sure @kiendang

kiendang commented on packages/syft/src/syft/service/action/action_graph.py:
```diff
@@ -79,7 +78,7 @@ def make_created_at(cls, v: Optional[DateTime]) -> DateTime:
 
     @staticmethod
-    def from_action(action: Action, credentials: SyftVerifyKey):
+    def from_action(action: Action, credentials: SyftVerifyKey) -> "NodeActionData":
```
This should be a `classmethod` returning `Self`. `return cls(...` instead of `return NodeActionData(...`.

khoaguin replied:
Thanks @kiendang. Done!

kiendang commented on packages/syft/src/syft/service/action/action_graph.py:
```diff
@@ -91,24 +90,26 @@ def from_action(action: Action, credentials: SyftVerifyKey):
-    def from_action_obj(action_obj: ActionObject, credentials: SyftVerifyKey):
+    def from_action_obj(
+        action_obj: ActionObject, credentials: SyftVerifyKey
+    ) -> "NodeActionData":
```
same here

kiendang commented on packages/syft/src/syft/service/action/action_graph.py:
```diff
@@ -148,7 +149,7 @@ def get(self, uid: Any) -> Any:
         raise NotImplementedError
 
-    def find_neighbors(self, uid: Any) -> List[Any]:
+    def find_neighbors(self, uid: Any) -> Optional[List[Any]]:
```
should this be
```suggestion
    def find_neighbors(self, uid: UID) -> Optional[List]:
```
? I saw another function with this signature on line 270.

also slight nitpick: could just be `Optional[List]` instead of `Optional[List[Any]]`.

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-dir

khoaguin committed:
[refactor] on fixing mypy issues for `service/queue/`
- comment out "--warn-unused-ignores" in precommit mypy
- small mypy fixes in `service/action`

Co-authored-by: Kien Dang <mail@kien.ai>

khoaguin committed:
[refactor] done fixing mypy issues for
`base_queue.py` and `queue.py`

khoaguin committed:
[refactor] fix cicular import error in queue.py

khoaguin committed:
delete unnecessary protocol json files

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-dir

khoaguin committed:
[refactor] done fixing mypy issues for `service/queue`

khoaguin committed:
[refactor] continue fixing mypy issues in

khoaguin committed:
[refactor] done fixing mypy issues for `service/code`

khoaguin committed:
[refactor] fix mypy issues in code related to user code and action service
- fix a bug in `test_code_accept_deny`

khoaguin committed:
[refactor] remove 'Execution denied' from Err: input policy is None

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-dir

khoaguin committed:
[refactor] continue fixing mypy issues for `syft/service`

kiendang commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -196,7 +196,9 @@ def syft_history_hash(self) -> int:
     @classmethod
-    def syft_function_action_from_kwargs_and_id(cls, kwargs, user_code_id):
+    def syft_function_action_from_kwargs_and_id(
+        cls, kwargs: Any, user_code_id: UID
```
should be

```suggestion
        cls, kwargs: Dict[str, Any], user_code_id: UID
```

`**kwargs: Any` but `kwargs: Dict[str, Any]`

kiendang commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -1260,7 +1266,7 @@ def obj_not_ready(
     @staticmethod
     def empty(
-        syft_internal_type: Type[Any] = NoneType,
+        syft_internal_type: Optional[Type[Any]] = NoneType,  # type: ignore[assignment]
```
```suggestion
        syft_internal_type: Optional[Type] = NoneType,  # type: ignore[assignment]
```

kiendang commented on packages/syft/src/syft/service/action/action_permissions.py:
```diff
@@ -46,7 +46,10 @@ def permission_string(self) -> str:
         if self.permission in COMPOUND_ACTION_PERMISSION:
             return f"{self.permission.name}"
         else:
-            return f"{self.credentials.verify}_{self.permission.name}"
+            if self.credentials is not None:
+                return f"{self.credentials.verify}_{self.permission.name}"
+            else:
+                return f"{self.permission.name}"
```
this can be unnested

```py
if self.credentials is not None:
    return f"{self.credentials.verify}_{self.permission.name}"
return f"{self.permission.name}"
```

kiendang commented on packages/syft/src/syft/service/code/user_code.py:
```diff
@@ -709,10 +733,10 @@ class SubmitUserCode(SyftObject):
 
     @property
-    def kwargs(self) -> List[str]:
+    def kwargs(self) -> Optional[dict[Any, Any]]:
```
check if this could be `Optional[dict[str, Any]]`

khoaguin replied:
I don't think so @kiendang since in the code there are already multiple places that it's defined as `Optional[Dict[Any, Any]]`
Also in line 410 of `user_code.py`, we use it with `[x.verify_key for x in self.input_policy_init_kwargs.keys()]` so it's probably not a string

kiendang replied:
ah I see

khoaguin committed:
[refactor] fix circular import errors
- remove some nested if else

Co-authored-by: Kien Dang <mail@kien.ai>

khoaguin committed:
[refactor] done fixing mypy issues in `syft/service/`

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-dir

khoaguin committed:
[refactor] remove unnecessary type checkings

khoaguin committed:
[bugfix] fix `test_code_accept_deny` not passed due to
wrong returned error message

khoaguin commented on packages/syft/src/syft/service/action/action_service.py:
```diff
@@ -291,6 +300,10 @@ def _user_code_execute(
 
         if not override_execution_permission:
             input_policy = code_item.input_policy
+            if input_policy is None:
```
I added an error handling here to make the `test_code_accept_deny` passed. I think it looks right, right? @koenvanderveen

kiendang commented on packages/syft/src/syft/util/telemetry.py:
```diff
@@ -85,4 +84,4 @@ def noop(func: T) -> T:
 
 
-instrument = setup_tracer()
+instrument: Callable = setup_tracer()
```
This one is tricky. `instrument` must have type `Callable[Concatenate[T, ..], T]`, not just `Callable`, otherwise we lose typing and auto-completion for any class that use the `@instrument` decorator. However, mypy does not seem to be able to infer the type for `instrument` if we just do `instrument = setup_tracer()` and if we set `instrument: Callable[Concatenate[T, ..], T] = setup_tracer()` it complains about wrong typing too. We'd have to modify the code a bit. I'll push a fix for this.

kiendang committed:
Fix instrument typing

kiendang committed:
Improve noop and instrument compatibility

Make _func_or_class positional-only in instrument
Make noop accept additional arbitrary args and kwargs

khoaguin committed:
[refactor] reenable checking `union-attr` mypy errors

khoaguin committed:
[refactor] using `cast(AbstractNode, context.node)` to
handle cases when context is AuthedServiceContext where
node is sure to present

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-dir

khoaguin committed:
[security] update `bandit` to `1.7.7`

khoaguin committed:
Merge branch 'dev' into fix-mypy-issues-service-dir

shubham3121 commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -659,18 +662,12 @@ def reload_cache(self):
                 blob_retrieval_object = blob_storage_read_method(
                     uid=self.syft_blob_storage_entry_id
                 )
-                if isinstance(blob_retrieval_object, SyftError):
```
Can we add this back ? and change the return typing for reload_cache to Optional[SyftErrror]

shubham3121 commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -775,25 +775,26 @@ def __check_action_data(cls, values: dict) -> dict:
-                    if hasattr(v, "_repr_markdown_")
-                    else v.__repr__()
-                )
+                if v is not None:
```
Any reason for checking v is not `None`. Even if v is None, then also v.__repr__() is a valid argument

shubham3121 replied:
Fixed by: 4961d8bd46eb13279bb7112cb5615f5f9ec8b6a5

shubham3121 commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -1710,11 +1719,12 @@ def _repr_markdown_(self) -> str:
-                    if hasattr(self.syft_action_data_cache, "_repr_markdown_")
-                    else self.syft_action_data_cache.__repr__()
-                )
+                if self.syft_action_data_cache is not None:
```
Why do we need to check for .syft_action_data_cache to be None ?

shubham3121 replied:
fixed by: 1e952fd0ef4e0f4f5be494ce530e7c2adacac7b9

shubham3121 commented on packages/syft/src/syft/service/action/action_service.py:
```diff
@@ -57,6 +60,8 @@ def __init__(self, store: ActionStore) -> None:
     def np_array(self, context: AuthedServiceContext, data: Any) -> Any:
         if not isinstance(data, np.ndarray):
             data = np.array(data)
+        if context.node is None:
```
can we cast here ?

shubham3121 commented on packages/syft/src/syft/service/action/action_service.py:
```diff
@@ -291,6 +300,10 @@ def _user_code_execute(
+            if input_policy is None:
+                if not code_item.output_policy_approved:
+                    return Err("Execution denied: Your code is waiting for approval")
+                return Err("Your code's input policy is None")
```
Can we change the message to say `No input poliicy defined for user code: {code_item.id}`

shubham3121 commented on packages/syft/src/syft/service/code/user_code.py:
```diff
@@ -425,22 +431,35 @@ def is_enclave_code(self) -> bool:
+    def input_owner_verify_keys(self) -> Optional[List[SyftVerifyKey]]:
+        if self.input_policy_init_kwargs is not None:
+            return [x.verify_key for x in self.input_policy_init_kwargs.keys()]
+        return None
```
we can skip returning None, its the default behaviour with no return types.

shubham3121 commented on packages/syft/src/syft/service/code/user_code.py:
```diff
@@ -571,11 +592,13 @@ def assets(self) -> List[Asset]:
-            for node_identity, uids in self.input_policy_init_kwargs.items()
-            if node_identity.node_name == api.node_name
-        )
+        if self.input_policy_init_kwargs is not None:
```
We're missing defining inputs if `self.input_policy_init_kwargs` is None

```python


inputs = () 
if self.input_policy_init_kwargs is not None:
    inputs = (
                uids
                for node_identity, uids in self.input_policy_init_kwargs.items()
                if node_identity.node_name == api.node_name
            )

```

shubham3121 commented on packages/syft/src/syft/service/code/user_code.py:
```diff
@@ -824,39 +852,40 @@ def _ephemeral_node_call(
-                res = ep_client.api.services.action.set(new_obj)
-                if isinstance(res, SyftError):
-                    return res
+        if self.input_policy_init_kwargs is not None:
```
can we type cast here `cast(self.input_policy_init_kwargs, dict)`

shubham3121 committed:
action_graph: fix type: ignore[call-arg] issue in action graph

shubham3121 committed:
add missing typing to TraceResult and _set_obj_location method

shubham3121 committed:
action_object: fix typing for syft_internal_type variable

shubham3121 committed:
action_object: fix _repr_markdown_ for ActionObject when syft_action_data_cache is None

shubham3121 committed:
action_object: fix storing of data repr when value is None

shubham3121 commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -914,7 +915,7 @@ def _syft_try_to_save_to_store(self, obj) -> None:
 
         if api is not None:
-            TraceResult.result += [action]
+            TraceResult.result += [action]  # type: ignore[unreachable]
```
fixed by: dd52fb249c90f35055c440ddd403dff512c6d848

shubham3121 commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -894,10 +895,10 @@ def _syft_try_to_save_to_store(self, obj) -> None:
 
         if api is not None:
-            obj._set_obj_location_(api.node_uid, api.signing_key.verify_key)
+            obj._set_obj_location_(api.node_uid, api.signing_key.verify_key)  # type: ignore[unreachable]
```
fixed by: dd52fb249c90f35055c440ddd403dff512c6d848

khoaguin committed:
[refactor] fixing mypy issues according
to Shubham's recommendations

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

shubham3121 committed:
fix typing in trace_decorator

khoaguin committed:
[refactor] using `cast(AbstractNode, context.node` to
pass type checking instead of returning `SyftError` or raise exceptions

shubham3121 committed:
remove some type ignores

shubham3121 committed:
remove type ignores in registry.py

shubham3121 committed:
worker_pool_service: cast context.node to AbstractNode

shubham3121 committed:
remove all occurences of type ignores from registry.py

shubham3121 committed:
fix return type of handle_api_call_with_unsigned_result

shubham3121 committed:
remove type ignore from user code
fix typing of object in job stash

shubham3121 committed:
cast context.node to AbstractNode in user and setting service

shubham3121 committed:
fix syntax for type casting of context.node to AbstractNode
- extract input and output init kwargs via getattr instead in syft_function decorator

shubham3121 reviewed (approved):
@khoaguin this is tremendous work !! 🦾 💯
